File: xurl-account-auth/scripts/oauth1-pin-auth.py
#!/usr/bin/env python3
import base64
import hashlib
import hmac
import os
import secrets
import subprocess
import sys
import time
import urllib.parse
import urllib.request

import yaml


def abort(message):
    raise SystemExit(message)


if len(sys.argv) != 3:
    abort("usage: uv run --with pyyaml python scripts/oauth1-pin-auth.py APP_NAME TARGET_USERNAME")

app_name = sys.argv[1]
target_username = sys.argv[2]
config_path = os.path.expanduser("~/.xurl")

with open(config_path, "r", encoding="utf-8") as file:
    config = yaml.safe_load(file)

app = config.get("apps", {}).get(app_name)
if not app:
    abort(f"app not found: {app_name}")

consumer_key = app.get("consumer_key") or app.get("client_id")
consumer_secret = app.get("consumer_secret") or app.get("client_secret")
if not consumer_key:
    abort(f"consumer key missing for app: {app_name}")
if not consumer_secret:
    abort(f"consumer secret missing for app: {app_name}")


def encode(value):
    return urllib.parse.quote(str(value), safe="~")


def oauth_header(method, url, params, token_secret=""):
    parameter_string = "&".join(f"{encode(key)}={encode(value)}" for key, value in sorted(params.items()))
    signature_base = "&".join([method.upper(), encode(url), encode(parameter_string)])
    signing_key = f"{encode(consumer_secret)}&{encode(token_secret)}"
    digest = hmac.new(signing_key.encode(), signature_base.encode(), hashlib.sha1).digest()
    signature = base64.b64encode(digest).decode()
    header_params = dict(params)
    header_params["oauth_signature"] = signature
    return "OAuth " + ", ".join(f'{encode(key)}="{encode(value)}"' for key, value in sorted(header_params.items()))


def post_form(url, oauth_params, token_secret=""):
    request = urllib.request.Request(url, method="POST")
    request.add_header("Authorization", oauth_header("POST", url, oauth_params, token_secret))
    with urllib.request.urlopen(request) as response:
        return response.read().decode()


request_token_url = "https://api.twitter.com/oauth/request_token"
request_params = {
    "oauth_callback": "oob",
    "oauth_consumer_key": consumer_key,
    "oauth_nonce": secrets.token_hex(16),
    "oauth_signature_method": "HMAC-SHA1",
    "oauth_timestamp": str(int(time.time())),
    "oauth_version": "1.0",
}

request_body = post_form(request_token_url, request_params)
request_data = dict(urllib.parse.parse_qsl(request_body))
oauth_token = request_data["oauth_token"]
oauth_token_secret = request_data["oauth_token_secret"]

subprocess.run(
    ["open", f"https://api.twitter.com/oauth/authorize?oauth_token={encode(oauth_token)}"],
    check=False,
)

pin = input("PIN: ").strip()
if not pin:
    abort("PIN missing")

access_token_url = "https://api.twitter.com/oauth/access_token"
access_params = {
    "oauth_consumer_key": consumer_key,
    "oauth_nonce": secrets.token_hex(16),
    "oauth_signature_method": "HMAC-SHA1",
    "oauth_timestamp": str(int(time.time())),
    "oauth_token": oauth_token,
    "oauth_verifier": pin,
    "oauth_version": "1.0",
}

access_body = post_form(access_token_url, access_params, oauth_token_secret)
access_data = dict(urllib.parse.parse_qsl(access_body))
authorized_username = access_data.get("screen_name")
if authorized_username != target_username:
    abort(f"authorized user is {authorized_username}, expected {target_username}")

app["consumer_key"] = consumer_key
app["consumer_secret"] = consumer_secret
app["oauth1_token"] = {
    "type": "oauth1",
    "oauth1": {
        "access_token": access_data["oauth_token"],
        "token_secret": access_data["oauth_token_secret"],
        "consumer_key": consumer_key,
        "consumer_secret": consumer_secret,
    },
}
config["default_app"] = app_name

with open(config_path, "w", encoding="utf-8") as file:
    yaml.safe_dump(config, file, allow_unicode=True, sort_keys=False)
os.chmod(config_path, 0o600)

print(f"saved oauth1 token for {authorized_username}")

File: xurl-account-auth/scripts/oauth1-pin-auth.rb
require "base64"
require "cgi"
require "net/http"
require "openssl"
require "securerandom"
require "uri"
require "yaml"

app_name = ARGV[0] || abort("usage: ruby scripts/oauth1-pin-auth.rb APP_NAME TARGET_USERNAME")
target_username = ARGV[1] || abort("usage: ruby scripts/oauth1-pin-auth.rb APP_NAME TARGET_USERNAME")

config_path = File.expand_path("~/.xurl")
config = YAML.load_file(config_path)
app = config.dig("apps", app_name)
abort "app not found: #{app_name}" unless app

consumer_key = app["consumer_key"] || app["client_id"]
consumer_secret = app["consumer_secret"] || app["client_secret"]
abort "consumer key missing for app: #{app_name}" if consumer_key.to_s.empty?
abort "consumer secret missing for app: #{app_name}" if consumer_secret.to_s.empty?

def encode(value)
  CGI.escape(value.to_s).gsub("+", "%20").gsub("%7E", "~")
end

def oauth_header(method, url, params, consumer_secret, token_secret = "")
  parameter_string = params.sort.map { |key, value| "#{encode(key)}=#{encode(value)}" }.join("&")
  signature_base = [method.upcase, encode(url), encode(parameter_string)].join("&")
  signing_key = "#{encode(consumer_secret)}&#{encode(token_secret)}"
  signature = Base64.strict_encode64(OpenSSL::HMAC.digest("sha1", signing_key, signature_base))
  header_params = params.merge("oauth_signature" => signature)
  "OAuth " + header_params.sort.map { |key, value| "#{encode(key)}=\"#{encode(value)}\"" }.join(", ")
end

request_token_url = "https://api.twitter.com/oauth/request_token"
request_params = {
  "oauth_callback" => "oob",
  "oauth_consumer_key" => consumer_key,
  "oauth_nonce" => SecureRandom.hex(16),
  "oauth_signature_method" => "HMAC-SHA1",
  "oauth_timestamp" => Time.now.to_i.to_s,
  "oauth_version" => "1.0"
}

request_uri = URI(request_token_url)
request = Net::HTTP::Post.new(request_uri)
request["Authorization"] = oauth_header("POST", request_token_url, request_params, consumer_secret)
response = Net::HTTP.start(request_uri.hostname, request_uri.port, use_ssl: true) { |http| http.request(request) }
abort "request token failed: #{response.code}" unless response.is_a?(Net::HTTPSuccess)

request_data = URI.decode_www_form(response.body).to_h
oauth_token = request_data.fetch("oauth_token")
oauth_token_secret = request_data.fetch("oauth_token_secret")
system("open", "https://api.twitter.com/oauth/authorize?oauth_token=#{encode(oauth_token)}")

print "PIN: "
pin = STDIN.gets&.strip
abort "PIN missing" if pin.to_s.empty?

access_token_url = "https://api.twitter.com/oauth/access_token"
access_params = {
  "oauth_consumer_key" => consumer_key,
  "oauth_nonce" => SecureRandom.hex(16),
  "oauth_signature_method" => "HMAC-SHA1",
  "oauth_timestamp" => Time.now.to_i.to_s,
  "oauth_token" => oauth_token,
  "oauth_verifier" => pin,
  "oauth_version" => "1.0"
}

access_uri = URI(access_token_url)
access_request = Net::HTTP::Post.new(access_uri)
access_request["Authorization"] = oauth_header("POST", access_token_url, access_params, consumer_secret, oauth_token_secret)
access_response = Net::HTTP.start(access_uri.hostname, access_uri.port, use_ssl: true) { |http| http.request(access_request) }
abort "access token failed: #{access_response.code}" unless access_response.is_a?(Net::HTTPSuccess)

access_data = URI.decode_www_form(access_response.body).to_h
authorized_username = access_data["screen_name"]
abort "authorized user is #{authorized_username}, expected #{target_username}" unless authorized_username == target_username

app["consumer_key"] = consumer_key
app["consumer_secret"] = consumer_secret
app["oauth1_token"] = {
  "type" => "oauth1",
  "oauth1" => {
    "access_token" => access_data.fetch("oauth_token"),
    "token_secret" => access_data.fetch("oauth_token_secret"),
    "consumer_key" => consumer_key,
    "consumer_secret" => consumer_secret
  }
}
config["default_app"] = app_name

File.write(config_path, YAML.dump(config))
File.chmod(0600, config_path)
puts "saved oauth1 token for #{authorized_username}"
